moto/config/models.py

```diff
@@ -411,13 +475,340 @@ def update(
         self.last_update_time = datetime2int(datetime.utcnow())
 
 
+class Scope(ConfigEmptyDictable):
+
+    """Defines resources that can trigger an evaluation for the rule.
+
+    Per boto3 documentation, Scope can be one of:
+    - one or more resource types,
+    - combo of one resource type and one resource ID,
+    - combo of tag key and value.
+
+    If no scope is specified, evaluations are trigged when any resource
+    in the recording group changes.
+    """
+
+    def __init__(
+        self,
+        compliance_resource_types=None,
+        tag_key=None,
+        tag_value=None,
+        compliance_resource_id=None,
+    ):
+        super().__init__(capitalize_start=True, capitalize_arn=False)
+        self.tags = None
+        if tag_key or tag_value:
+            if tag_value and not tag_key:
+                raise InvalidParameterValueException(
+                    "Tag key should not be empty when tag value is provided in scope"
+                )
+            if tag_key and len(tag_key) > 128:
+                raise TagKeyTooBig(tag_key, "ConfigRule.Scope.TagKey")
+            if tag_value and len(tag_value) > 256:
+                raise TagValueTooBig(tag_value, "ConfigRule.Scope.TagValue")
+            self.tags = {tag_key: tag_value}
+
+        # Can't use more than one combo to specify scope - either tags,
+        # resource types, or resource id and resource type.
+        if self.tags and (compliance_resource_types or compliance_resource_id):
+            raise InvalidParameterValueException(
+                "Scope cannot be applied to both resource and tag"
+            )
+
+        if compliance_resource_id and len(compliance_resource_types) != 1:
+            raise InvalidParameterValueException(
+                "A single resourceType should be provided when resourceId "
+                "is provided in scope"
+            )
+        self.compliance_resource_types = compliance_resource_types
+        self.compliance_resource_id = compliance_resource_id
+
+
+class SourceDetail(ConfigEmptyDictable):
+
+    """Source and type of event triggering AWS Config resource evaluation.
+
+    Applies only to customer rules.
+    """
+
+    MESSAGE_TYPES = [
+        "ConfigurationItemChangeNotification",
+        "OversizedConfigurationItemChangeNotification",
+        "ScheduleNotification",
+        "ConfigurationSnapshotDeliveryCompleted",
+    ]
+    DEFAULT_FREQUENCY = "TwentyFour_Hours"
+    FREQUENCY_TYPES = [
+        "TwentyFour_Hours",
+        "Twelve_Hours",
+        "Three_Hours",
+        "Six_Hours",
+        "One_Hour",
+    ]
+    EVENT_SOURCES = ["aws.config"]
+
+    def __init__(
+        self, event_source, message_type, max_execution_frequency=DEFAULT_FREQUENCY
+    ):
+        super().__init__(capitalize_start=True, capitalize_arn=False)
+        if not event_source:
+            # boto3 doesn't have a specific error if this field is missing.
+            raise ParamValidationError(
```

Comment:
ParamValidationErrors are raised by boto3 itself, before the request ever gets send to Moto - it's like an initial validation.
So we don't need these

Reply:
Thank you.  I should have known better.

Reply:
All good - I prefer too many tests over too few!

Reply:
Turns out I was wrong about this - boto does not intercept this particular error, so catching it manually does make sense.
Incidentally, it's also the cause of the ServerMode error that we're seeing now.

The Moto internals can gracefully deal with Exceptions - as long as it extends a specific werkzeug HTTPException.
Any other exceptions, it doesn't know how to deal with those, so it throws a generic 500-exception instead of the custom crafted exception that Boto3 expects.

The solution would be to create a custom Exception for this, and ensure that it extends the JsonRESTError-class, like all our custom exceptions.

```
class WhateverNameMakesSense(JsonRESTError):
    code = 400

    def __init__(self):
        super().__init__(
            "ParamValidation",
            "Missing required parameter in ConfigRule.SourceDetails: 'EventSource'",
        )
```

The best way to verify that Moto throws the correct error, is to always follow this format in the unit tests:
```
with pytest.raises(ClientError) as exc:
        client.put_config_rule(ConfigRule=custom_rule)
err = exc.value.response["Error"]
err["Message"].should.equal("Missing required parameter in ConfigRule.SourceDetails: 'EventSource'")
err["Code"].should.equal("ParamValidation")
```
This test will only pass if the error is returned in the correct format.

Reply:
Is it ok if I use `assert` versus the `sure` library ("should.equal")?

I'm also having trouble if I use your suggested code and I think it's because the error isn't really a ClientError.  The error I'm seeing is as follows (using your suggested code):

```
{'Message': '{\n    
"message": "Missing required parameter in ConfigRule.SourceDetails: "EventSource"",\n    
"__type": "ParamValidationError"\n
}', 
'Code': '400'}
```

The expected format of a ClientError is more like this:

```
{'Message': "Value 'foo' at 'configRule.source.sourceDetails.eventSource'
failed to satisfy constraint: Member must satisfy enum value set: {aws.config}", 
'Code': 'ValidationException'}
```

I can see there are other error types in `moto.core.exceptions.py`, but I'm concerned that the users of this API will expect a ClientError.  I can convert the incoming message to a ClientError type, but I'm not sure if you'd be ok with that.  I'll give that a try so I continue adding some missing unit tests, but I'm happy to change it based on your feedback/decision.

Reply:
Make sure that the error msg is valid JSON, i.e. small quotes around the 'EventSource'.
If the message can't be used directly as a JSON key, that's when you see the weird format.

Reply:
Damn you're good!  That was the problem.  Thank you ...

Reply:
> 
> 
> Damn you're good! That was the problem. Thank you ...

Yeah.. I already ran into the same issue when I was debugging. I just forgot to mention it 🙁 

> 
> 
> Is it ok if I use `assert` versus the `sure` library ("should.equal")?

Yeah, that's absolutely fine. I prefer `should.equal` because it reads better IMO, but it's very much a personal preference, so we don't really enforce any style when it comes to assertions.